### PR TITLE
Faster Tree folds

### LINKED ISSUE #878: Poor performance of Foldable for Data.Tree.Tree
Folds on `Tree` seem to perform very poorly.

```hs
{-# LANGUAGE BangPatterns #-}
import Data.Foldable
import Data.Monoid
import Data.Tree
import Test.Tasty.Bench

main :: IO ()
main = defaultMain
    [ env (pure tree_) $ \t -> bgroup ""
        [ bench "sum"              $ whnf sum t
        , bench "foldMap'Tree Sum" $ whnf (foldMap'Tree Sum) t
        ]
    ]
  where
    lim = 1000 * 1000
    tree_ = unfoldTree (\x -> (x, filter (<=lim) [2*x, 2*x+1])) 1 :: Tree Int

foldMap'Tree :: Monoid m => (a -> m) -> Tree a -> m
foldMap'Tree f = go mempty
  where go !z (Node x ts) = foldl' go (z <> f x) ts
```
```
    sum:              OK (1.82s)
      249  ms ±  15 ms
    foldMap'Tree Sum: OK (0.31s)
      10.2 ms ± 717 μs
```

### PR DESCRIPTION
**Pull request by meooow25**
Add some `Foldable` functions for `Data.Tree.Tree`.

* `foldMap`, `null`: Already exist
* `foldr`, `foldl'`: Added, basic functions
* `fold`, `elem`: Added, default definitions but `INLINABLE` helps
* `foldr1`, `foldl1`, `maximum`, `minimum`: Added, default definitions introduce `Maybe`s since these can be partial, we can avoid all that since trees are non-empty.
* `foldMap'`, `toList`, `length`, `sum`, `product`: Not added, defaults work well enough
* `foldr'`, `foldl`: Not added or benchmarked, unlikely to be used(?)

Benchmarks:

```
  fold:    OK (0.25s)
    15.5 μs ± 962 ns, 85% less than baseline
  foldMap: OK (0.14s)
    18.0 μs ± 1.7 μs, 82% less than baseline
  foldr:   OK (0.18s)
    21.8 μs ± 1.5 μs, 93% less than baseline
  foldl':  OK (0.30s)
    18.1 μs ± 878 ns, 95% less than baseline
  foldr1:  OK (0.16s)
    19.6 μs ± 1.5 μs, 94% less than baseline
  foldl1:  OK (0.30s)
    18.0 μs ± 1.2 μs, 95% less than baseline
  toList:  OK (0.17s)
    39.9 μs ± 2.9 μs, 31% less than baseline
  elem:    OK (0.27s)
    16.3 μs ± 777 ns, 47% less than baseline
  maximum: OK (0.16s)
    19.6 μs ± 1.5 μs, 94% less than baseline
  sum:     OK (0.30s)
    17.9 μs ± 671 ns, 95% less than baseline
```

Fixes #878.

### CODE REVIEW

**Comment by treeowl** [comment]
95% less than? So only 5% of the earlier time? That sounds ... A bit suspicious. Is it right?

**Comment by meooow25** [comment]
Unfortunately yes. Currently it seems to be ~20x slow for something like `sum`. I saw a similar difference when I reported the issue, and here's a comparison vs summing a list of the same size.
```
  sum Tree Int old: OK (0.21s)
    393  μs ±  23 μs
  sum [Int]:        OK (0.23s)
    13.9 μs ± 800 ns
```
```
  sum Tree Int new: OK (0.15s)
    18.2 μs ± 1.4 μs
```

**Comment by treeowl** [comment]
It's apparent that the trees in the test aren't big enough. I don't love comparing microseconds. Could you possibly bump those sizes up some and see what things look like?

**Comment by meooow25** [comment]
What size would you prefer? I chose `2^12` because that's what other container tests are using.

**Comment by treeowl** [comment]
I don't know, really. Ideally we'd have a few sizes, but for now aim for half a second each, unless that makes the benchmarking take too long.

**Comment by meooow25** [comment]
Half a second would need something _very_ large. I've updated it to `2^20`, a little over a million.
```
  fold:    OK (0.76s)
    8.45 ms ± 366 μs, 76% less than baseline
  foldMap: OK (0.40s)
    8.89 ms ± 683 μs, 73% less than baseline
  foldr:   OK (0.42s)
    10.7 ms ± 753 μs, 94% less than baseline
  foldl':  OK (0.42s)
    10.5 ms ± 949 μs, 95% less than baseline
  foldr1:  OK (0.26s)
    10.0 ms ± 989 μs, 95% less than baseline
  foldl1:  OK (0.42s)
    10.5 ms ± 748 μs, 96% less than baseline
  toList:  OK (0.35s)
    30.0 ms ± 1.8 ms,  9% less than baseline
  elem:    OK (0.43s)
    10.3 ms ± 701 μs, 28% less than baseline
  maximum: OK (0.42s)
    10.5 ms ± 788 μs, 96% less than baseline
  sum:     OK (0.26s)
    10.5 ms ± 725 μs, 96% less than baseline
```

File: containers/src/Data/Tree.hs
Comment (treeowl) [inline containers/src/Data/Tree.hs:195]: Is there a reason not to tuck `z` under the lambda?

I imagine you pass `z` through the recursion because that performs better than doing the obvious thing. Have you verified that it's better? If so, a note explaining that would be helpful.

Comment (meooow25) [reply containers/src/Data/Tree.hs]: I assume you mean "under the `go`"? Yes I had tried a few variations and found it performs well.
Added a comment.

Looking at the core and it does seem to generate better output. I wouldn't be able to explain why though.
For the record, on the `(&&)` benchmark
```hs
go (Node x ts) = f x . foldr ((.) . go) id ts
```
generates core like
```
Rec {
-- RHS size: {terms: 20, types: 19, coercions: 0, joins: 0/0}
go1 :: [Tree Bool] -> Bool -> Bool
[GblId, Arity=2, Str=<1L><ML>, Unf=OtherCon []]
go1
  = \ (ds :: [Tree Bool]) (eta :: Bool) ->
      case ds of {
        [] -> eta;
        : y ys ->
          case y of { Node ww1 ww2 ->
          case ww1 of {
            False -> False;
            True -> go1 ww2 (go1 ys eta)
          }
          }
      }
end Rec }
```
which is ideal, and
```hs
go (Node x ts) z' = f x (foldr go z' ts)
```
generates
```
Rec {
-- RHS size: {terms: 25, types: 27, coercions: 0, joins: 0/1}
$wgo3 :: Bool -> [Tree Bool] -> Bool -> Bool
[GblId, Arity=3, Str=<1L><ML><L>, Unf=OtherCon []]
$wgo3
  = \ (ww :: Bool) (ww1 :: [Tree Bool]) (w :: Bool) ->
      case ww of {
        False -> False;
        True ->
          letrec {
            go1 [Occ=LoopBreaker, Dmd=SCS(L)] :: [Tree Bool] -> Bool
            [LclId, Arity=1, Str=<1L>, Unf=OtherCon []]
            go1
              = \ (ds :: [Tree Bool]) ->
                  case ds of {
                    [] -> w;
                    : y ys -> case y of { Node ww3 ww4 -> $wgo3 ww3 ww4 (go1 ys) }
                  }; } in
          go1 ww1
      }
end Rec }
```

File: containers/src/Data/Tree.hs
Comment (phadej) [inline containers/src/Data/Tree.hs:197]: this is a unjustified claim. Microbenchmarks are not enough to establish that. Given that `foldr` is marked `INLINE`, it's optimized at each use case

I'd go for simpler version if the difference in benchmarks small, as GHC optimizer is unrelieable, and perfectly should be tuned to make "normal" code run smarter. I.e. in perfect future, the simple version would optimize as well.

If the difference is big, than it's should be reported to GHC upstream.

I'm also not a fan of having `INLINE` for "performance". It's not sustainable, causes code growth (makes microbenchmarks look better, but real code worse). `INLINEABLE` should be enough for specialization.

File: containers/src/Data/Tree.hs
Comment (phadej) [inline containers/src/Data/Tree.hs:216]: You should also implement `sum` and `product`, as you on this.

And `foldMap'` for `base >=4.13`.

File: containers/src/Data/Tree.hs
Comment (phadej) [inline containers/src/Data/Tree.hs:224]: GHC will barf on `INLINE` on recursive functions. I.e. it does nothing. `INLINEABLE` for specialization is enough.

File: containers/src/Data/Tree.hs
Comment (phadej) [reply containers/src/Data/Tree.hs:197]: Forgot to mention: `foldr` doesn't need to be "fast", it meant to be lazy (e.g. for lists), and can be for `Tree` as well. 

Benchmarking something like `head (toList t)` is a good benchmark for `foldr`. It should be independent of `t` size.

Comment (meooow25) [reply containers/src/Data/Tree.hs:197]: > this is a unjustified claim. Microbenchmarks are not enough to establish that. Given that foldr is marked INLINE, it's optimized at each use case

Fair point, but we have to pick one implementation, and benchmarks help us decide. If you would prefer the comment said something else I'm open to suggestions.

> I'd go for simpler version if the difference in benchmarks small

Current
```
  foldr: OK (0.36s)
    11.1 ms ± 840 μs
```
Simpler
```
  foldr: OK (0.26s)
    18.2 ms ± 1.5 ms
```

> If the difference is big, than it's should be reported to GHC upstream.

That's reasonable, I can open an issue.

> I'm also not a fan of having INLINE for "performance".

It seems necessary, `INLINABLE` doesn't always work as well. Every `foldr` in containers (`Set`, `Map`, `IntSet`, `IntMap`, `FingerTree` for `Sequence`) is `INLINE` for performance.

> Forgot to mention: foldr doesn't need to be "fast", it meant to be lazy

Why not both?

> Benchmarking something like head (toList t) is a good benchmark for foldr. 

`head (toList t)` would finish instantaneously. I don't see the point of that. But I'm happy to add/replace foldr benchmarks with other use cases. There is already `toList` and `&&` is the best I could think of for something reducing to a small value.

File: containers/src/Data/Tree.hs
Comment (meooow25) [reply containers/src/Data/Tree.hs:216]: As I mentioned in the description `sum`, `product` and `foldMap'` are working well by default. The only thing we gain by implementing these is saving one operation with the initial value. Which I can implement, sure, but is that what you're asking for?

File: containers/src/Data/Tree.hs
Comment (meooow25) [reply containers/src/Data/Tree.hs:224]: This is not a recursive function.

File: containers/src/Data/Tree.hs
Comment (phadej) [reply containers/src/Data/Tree.hs:216]: description is not visible in the source nor version control. Why `elem` is expanded, but `foldMap'` isn't? Write the note.

I also doubt that `sum` works well on old GHCs, before it was changed to use `foldMap'`.

File: containers/src/Data/Tree.hs
Comment (phadej) [reply containers/src/Data/Tree.hs:224]: Ah indeed!

File: containers/src/Data/Tree.hs
Comment (phadej) [reply containers/src/Data/Tree.hs:197]: What's the difference between current and simpler implementation if you don't INLINE them.

> It seems necessary, INLINABLE doesn't always work as well. Every foldr in containers (Set, Map, IntSet, IntMap, FingerTree for Sequence) is INLINE for performance.

And that is bad. Don't make things worse.

> head (toList t) would finish instantaneously.

It's a correctness test. `take n (toList infinitelySwawningTree)` should always work.
e.g. `let t = Node 'x' (repeat t)` or just `t = Node 'x' [t]`.

Comment (treeowl) [reply containers/src/Data/Tree.hs:197]: @phadej, inclining a fold for a simple structure generally shouldn't increase code size much if at all. What's the problem you see with it?

Comment (treeowl) [reply containers/src/Data/Tree.hs:197]: I tend to want two benchmarks for `foldr`, to reflect two use patterns:

1. `length . foldr (:) []`, for the lazy production scenario.
2. Some strict left fold with abort for other common uses. It's too early in the morning for me to come up with a nice one.

Comment (phadej) [reply containers/src/Data/Tree.hs:197]: > inlining a fold for a simple structure generally shouldn't increase code size much if at all. What's the problem you see with it?

The little streams make the big rivers.

YMMV.

File: containers/src/Data/Tree.hs
Comment (meooow25) [reply containers/src/Data/Tree.hs:219]: Done, added a note and implemented `sum` and `product`.

File: containers/src/Data/Tree.hs
Comment (meooow25) [reply containers/src/Data/Tree.hs]: Added the `length . foldr (:) []` benchmark.

> What's the difference between current and simpler implementation if you don't INLINE them.

|`foldr (&&) True`|`INLINE`|`INLINABLE`|Nothing|
|---|---|---|---|
|Current|11.0 ms ± 546 μs|23.3 ms ± 1.6 ms|23.2 ms ± 1.5 ms|
|Simple|18.2 ms ± 1.4 ms|18.3 ms ± 1.5 ms|30.0 ms ± 1.4 ms|

|`length . foldr (:) []`|`INLINE`|`INLINABLE`|Nothing|
|---|---|---|---|
|Current|17.5 ms ± 1.7 ms|21.5 ms ± 1.6 ms|21.4 ms ± 1.8 ms|
|Simple|31.3 ms ± 1.6 ms|30.0 ms ± 2.4 ms|35.6 ms ± 2.8 ms|

I'm strongly in favor of keeping the current implementation.

> It's a correctness test.

Well then that's a test, not a benchmark. Added.

Comment (meooow25) [reply containers/src/Data/Tree.hs]: @treeowl what do you think?
And let me know if want me to add an example of 2.

Comment (treeowl) [reply containers/src/Data/Tree.hs]: I definitely want an example of a left fold with abort. Something like

```haskell
foldr go stop t (0 :: Int)
  where
    go x r acc
      | acc > some_constant = acc
      | otherwise = r (acc + x)
```

I think that's an important `foldr` pattern, and we want to make sure it's optimized reasonably well.

Comment (meooow25) [reply containers/src/Data/Tree.hs]: Added.

**Comment by meooow25** [comment]
Added benchmarks using a line tree as an example of a deep tree, since a binary tree is rather shallow. Also added a smaller size of 1000.

And I opened a GHC issue about the fold function: https://gitlab.haskell.org/ghc/ghc/-/issues/22659

<details>
<summary>New benchmarks</summary>
<pre>
  fold
    bin,n=1000:     OK (0.96s)
      4.20 μs ± 347 ns
    bin,n=1000000:  OK (0.58s)
      8.86 ms ± 820 μs
    line,n=1000:    OK (1.70s)
      5.30 μs ± 442 ns
    line,n=1000000: OK (0.56s)
      14.8 ms ± 802 μs
  foldMap
    bin,n=1000:     OK (1.74s)
      4.58 μs ± 338 ns
    bin,n=1000000:  OK (0.45s)
      8.93 ms ± 869 μs
    line,n=1000:    OK (1.77s)
      5.79 μs ± 367 ns
    line,n=1000000: OK (0.58s)
      15.2 ms ± 1.1 ms
  foldr_1
    bin,n=1000:     OK (1.75s)
      5.51 μs ± 331 ns
    bin,n=1000000:  OK (0.59s)
      10.7 ms ± 723 μs
    line,n=1000:    OK (1.58s)
      7.34 μs ± 667 ns
    line,n=1000000: OK (3.56s)
      46.1 ms ± 3.3 ms
  foldr_2
    bin,n=1000:     OK (1.59s)
      8.42 μs ± 716 ns
    bin,n=1000000:  OK (0.69s)
      17.0 ms ± 1.2 ms
    line,n=1000:    OK (1.42s)
      9.31 μs ± 757 ns
    line,n=1000000: OK (0.60s)
      34.6 ms ± 2.5 ms
  foldr_3
    bin,n=1000:     OK (1.63s)
      7.88 μs ± 674 ns
    bin,n=1000000:  OK (0.66s)
      15.0 ms ± 904 μs
    line,n=1000:    OK (1.78s)
      7.25 μs ± 365 ns
    line,n=1000000: OK (0.43s)
      28.5 ms ± 1.8 ms
  foldl'
    bin,n=1000:     OK (1.09s)
      4.43 μs ± 335 ns
    bin,n=1000000:  OK (0.61s)
      10.5 ms ± 753 μs
    line,n=1000:    OK (1.12s)
      5.81 μs ± 439 ns
    line,n=1000000: OK (0.78s)
      20.8 ms ± 1.1 ms
  foldr1
    bin,n=1000:     OK (2.08s)
      5.54 μs ± 207 ns
    bin,n=1000000:  OK (0.60s)
      10.1 ms ± 989 μs
    line,n=1000:    OK (1.78s)
      6.69 μs ± 443 ns
    line,n=1000000: OK (0.37s)
      19.6 ms ± 1.4 ms
  foldl1
    bin,n=1000:     OK (1.73s)
      4.35 μs ± 377 ns
    bin,n=1000000:  OK (0.87s)
      10.3 ms ± 419 μs
    line,n=1000:    OK (1.14s)
      6.27 μs ± 609 ns
    line,n=1000000: OK (2.51s)
      30.6 ms ± 1.6 ms
  toList
    bin,n=1000:     OK (1.61s)
      9.13 μs ± 761 ns
    bin,n=1000000:  OK (0.54s)
      29.4 ms ± 1.6 ms
    line,n=1000:    OK (1.61s)
      10.7 μs ± 814 ns
    line,n=1000000: OK (0.38s)
      49.4 ms ± 4.8 ms
  elem
    bin,n=1000:     OK (2.00s)
      4.48 μs ± 288 ns
    bin,n=1000000:  OK (0.88s)
      10.5 ms ± 440 μs
    line,n=1000:    OK (1.02s)
      5.50 μs ± 383 ns
    line,n=1000000: OK (0.37s)
      20.8 ms ± 1.7 ms
  maximum
    bin,n=1000:     OK (1.74s)
      4.52 μs ± 387 ns
    bin,n=1000000:  OK (0.37s)
      10.5 ms ± 974 μs
    line,n=1000:    OK (1.14s)
      6.16 μs ± 342 ns
    line,n=1000000: OK (0.79s)
      20.9 ms ± 722 μs
  sum
    bin,n=1000:     OK (1.99s)
      4.47 μs ± 285 ns
    bin,n=1000000:  OK (0.89s)
      10.3 ms ± 392 μs
    line,n=1000:    OK (1.15s)
      6.36 μs ± 388 ns
    line,n=1000000: OK (0.79s)
      21.0 ms ± 1.7 ms
</pre>
</details>

File: containers/src/Data/Tree.hs
Comment (phadej) [inline containers/src/Data/Tree.hs:198]: In https://gitlab.haskell.org/ghc/ghc/-/issues/22659#note_469510 SPJ identifies why this is faster. And sebastian suggest that this variant may be slower for deep trees (i.e. something like `Node 1 $ singleton $ Node 2 $ singleton $ ...`)

I suggest:
- Add benchmarks for "simple" foldr as well, so the difference can be checked later as well
- Add benchmark for deep trees
- Rewrite this using two mutually recursive functions as SPJ shows. That implementation is less magical, relies less on optimizer, and is easy to explain why it behaves better.

Comment (meooow25) [reply containers/src/Data/Tree.hs:198]: > sebastian suggest that this variant may be slower for deep trees

That is not what Sebastian is saying. If I understand the comment correctly, Sebastian says that if we had two mutually recursive functions after optimization, we would allocate more. The current code optimizes to a single function.

> Add benchmarks for "simple" foldr as well, so the difference can be checked later as well

I don't see the point of this. Anyone running benchmarks would have to filter out the noise of a benchmark that is not live code. If anyone wants to test the simple version that's easy, they just need to comment out this line and uncomment the definition below.

> Add benchmark for deep trees

I have, please read the comment above.

> Rewrite this using two mutually recursive functions as SPJ shows. That implementation is less magical, relies less on the optimizer, and is easy to explain why it behaves better.

The second function out of SPJ's two mutually recursive functions manually case matches on the list and recurses. We don't need to write code like that when we have folds to do the job for us. And if we use foldr, that is the same as the simple definition, only broken into two, and it optimizes to the same inefficient code.

I honestly don't understand your dislike for this implementation.

Comment (phadej) [reply containers/src/Data/Tree.hs:198]: At the very least expand

```haskell
(.) . go
```

it's a pointless point-free form.

Comment (phadej) [reply containers/src/Data/Tree.hs:198]: > I don't see the point of this. Anyone running benchmarks would have to filter out the noise of a benchmark that is not live code. If anyone wants to test the simple version that's easy, they just need to comment out this line and uncomment the definition below.

I disagree about "easy". But as I'm not a maintainer, I cannot require you to add it either.

Comment (meooow25) [reply containers/src/Data/Tree.hs]: > At the very least expand `(.) . go`

Sure thing, done.

Comment (phadej) [reply containers/src/Data/Tree.hs]: Ok now I undertstand what happens.

```haskell
foldr f z = \(Node x ts) -> f x (foldrForest f ts z)

foldrForest :: (a -> b -> b) -> [Tree a] -> b -> b
foldrForest f = go where 
  go []               z = z
  go (Node x xs : ys) z = f x (go xs (go ys z))
{-# INLINE foldrForest #-}
```

I find this easier to understand that one written using list's `foldr`. YMMV

Comment (meooow25) [reply containers/src/Data/Tree.hs]: Yes, your function seems to end up in the same form as the current implementation. But I do not find the current one hard to understand. Of course, this is subjective. When something can be written as a fold, I would prefer to write it as a fold instead of explicit recursion unless there is a strong reason to do otherwise.

@treeowl do you have a preference here?
Otherwise, if there's nothing else to add, can we close this?

**Comment by treeowl** [comment]
Sorry, I got distracted with some lens stuff. I like folds, if they're clear and fast. I feel similarly about explicit recursion. Whatever's nicest in context.

**Comment by meooow25** [comment]
I'm happy with the current state of things. If you agree, we can merge it.

**Comment by meooow25** [comment]
Sorry, I feel like I'm missing something. What's the hold up here? Please tell me clearly.

**Comment by treeowl** [comment]
I need to make the time for a final review. I will try to do that tomorrow.

**Comment by meooow25** [comment]
That's understandable, thank you.

**Comment by meooow25** [comment]
Hi, did you get time to review it?

**Comment by treeowl** [comment]
It will happen tonight without fail!

File: containers-tests/tests/tree-properties.hs
Comment (treeowl) [inline containers-tests/tests/tree-properties.hs:138]: This test is really just about termination, right? Anything else it has to say will be captured by finite tests. Wouldn't it be better to use something like this to get more infinite tree shapes?

```haskell
prop_foldr_infinite n = forAll genny (\genny -> length (take n (foldr (:) [] t)) === n)
  where
    genny :: Gen (Tree ())
    genny = Node () <$> oneof [listOf genny, infiniteListOf genny]
```

File: containers/src/Data/Tree.hs
Comment (treeowl) [inline containers/src/Data/Tree.hs:229]: I don't see tests for any of these methods. Could you add a sum test and a minimum test?

File: containers/src/Data/Tree.hs
Comment (treeowl) [inline containers/src/Data/Tree.hs:477]: I don't understand what makes `fold` and `elem` different from `foldMap'` in this regard. Are there issues specializing the first two but not the last? If so, do you have any guess why?

**Review by treeowl — CHANGES_REQUESTED** [review]
I'm sorry for the long delay. I've marked up some questions and requests.

**Comment by treeowl** [comment]
Did you see my review?

**Comment by treeowl** [comment]
For `foldl1` and `foldr1`, I would recommend testing with the operation of a free magma.

```haskell
data Magma a
  = Inj a
  | Magma a :* Magma a
  deriving (Eq, Show)
```

File: containers/src/Data/Tree.hs
Comment (meooow25) [reply containers/src/Data/Tree.hs:196]: Looks like it works, done.

File: containers/src/Data/Tree.hs
Comment (meooow25) [reply containers/src/Data/Tree.hs]: In that case I think we can skip it. We provide the simple function below and the discussion is a bit noisy and not too important.

File: containers/src/Data/Tree.hs
Comment (meooow25) [reply containers/src/Data/Tree.hs:480]: The default definition `foldMap'` is perhaps small enough for GHC to inline without any help. But I don't know for sure.

File: containers-tests/tests/tree-properties.hs
Comment (meooow25) [reply containers-tests/tests/tree-properties.hs]: Done.

File: containers-tests/tests/tree-properties.hs
Comment (meooow25) [reply containers-tests/tests/tree-properties.hs:137]: Now that foldMap is defined as foldMapDefault, might be a little odd to test this?

File: containers-tests/tests/tree-properties.hs
Comment (meooow25) [reply containers-tests/tests/tree-properties.hs]: Sure, done. This generator can generate a finite tree though (`listOf genny` can be null), so I added an infinite of these at the top level.

File: containers/src/Data/Tree.hs
Comment (meooow25) [reply containers/src/Data/Tree.hs:232]: Added for all four.

**Comment by meooow25** [comment]
> Did you see my review?

Hi yes, was going through your comments.

> For `foldl1` and `foldr1`, I would recommend testing with the operation of a free magma.

Thanks, I was thinking the same but I wasn't sure what to name it. And in fact `foldr1` did give different results for a non-associative function, so thanks for catching this. Fixed now.

File: containers/src/Data/Tree.hs
Comment (meooow25) [reply containers/src/Data/Tree.hs:198]: Right, I don't think it would be bad to do that, but all `foldrs` in containers are defined as `foldr f z` (even when z can be eta-reduced, not sure why), so I've kept it consistent.
I also noticed this brought up in [GHC !5259](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/5259) for `foldl'`. But it seems that it was not definitively addressed, and in the end they went with `foldl' f z0`.

File: containers/src/Data/Tree.hs
Comment (treeowl) [reply containers/src/Data/Tree.hs:480]: Bleh. GHC inlining pragmas are such a dark art. I hate it.

File: containers-tests/tests/tree-properties.hs
Comment (treeowl) [reply containers-tests/tests/tree-properties.hs]: Generating a finite tree sometimes is fine. We're going to be generating a *lot* of trees. I'd keep it simple.

Comment (treeowl) [reply containers-tests/tests/tree-properties.hs]: Oh, but I forgot that it should therefore be `length ... <= n`!

File: containers/src/Data/Tree.hs
Comment (treeowl) [reply containers/src/Data/Tree.hs:198]: OK. There might be/have been a reason. The secrets of the GHC inliner....

File: containers-tests/tests/tree-properties.hs
Comment (treeowl) [reply containers-tests/tests/tree-properties.hs:137]: I know it's odd, but that'll catch a mistake if someone later decides to go back to implementing it manually.

File: containers-tests/tests/tree-properties.hs
Comment (meooow25) [reply containers-tests/tests/tree-properties.hs]: Oh that's what you meant, done.

File: containers-tests/tests/tree-properties.hs
Comment (meooow25) [reply containers-tests/tests/tree-properties.hs]: Okay, done.

File: containers-tests/tests/tree-properties.hs
Comment (treeowl) [inline containers-tests/tests/tree-properties.hs:134]: I'm sorry I didn't notice this earlier, but in this test suite (and most of the rest in the package) we tend to use `Test.QuickCheck.Poly` where applicable. So for this test we should have

```haskell
prop_toList :: Tree A -> Property
```

The same goes for the rest. Yes, it's a bit less efficient (`A`, `B`, `C`, etc., are actually `Integer`), but clear communication is more important than constant factors for these tests. For `prop_maximum`, ..., `prop_product`, you'll need to use `OrdA`, etc., instead (they have `Ord` and `Num` instances).

Comment (meooow25) [reply containers-tests/tests/tree-properties.hs]: Ah makes sense, done.

**Review by treeowl — APPROVED** [review]
Whew! Let's do it.

**Comment by treeowl** [comment]
Could you please squash this into one or more logical commits?

**Comment by meooow25** [comment]
Squashed them all!

**Pull request merged**
